app/src/main/java/com/UserDatabase/UserDatabase.java

```diff
@@ -0,0 +1,264 @@
+            int ch = fileInputStream.read();
+
+            while (ch != -1) {
+                databaseString.append((char) ch);
+                ch = fileInputStream.read();
+            }
+        } catch (IOException e) {
+            Log.e(LogTag, "Cannot load database");
+            e.printStackTrace();
+            return;
+        }
+
+        // Deserialize database string
+        Gson gson = new Gson();
+        JsonObject databaseJson = gson.fromJson(databaseString.toString(), JsonObject.class);
+
+        String loadedId = databaseJson.get("Id").getAsString();
+        int loadedVectorLength = databaseJson.get("VectorLength").getAsInt();
+        String serializedUserRecords = databaseJson.get("UserRecords").getAsString();
+
+        // Validate database
+        // TODO: Later it might be better to throw exception here (in order to display dialog box or sth)
+        assertEquals("Wrong type of database", Id, loadedId);
+        assertEquals("Wrong size of database", VectorLength, loadedVectorLength);
+
+        // Load users records
+        usersRecords.clear();
+        Type userDatabaseType = new TypeToken<Map<String, float[]>>() {}.getType();
+        usersRecords = gson.fromJson(serializedUserRecords, userDatabaseType);
+
+        Log.d(LogTag, "Database file successfully loaded");
+    }
+
+    /**
+     * Serialize user database.
+     *
+     */
+    public void saveDatabase() {
+
+         // Serialize userRecords to Json
+        Gson gson = new Gson();
+        String serializedUserRecords = gson.toJson(usersRecords);
+
+        // Complete Json object with database header data
+        JsonObject databaseJson = new JsonObject();
+        databaseJson.addProperty("Id", Id);
+        databaseJson.addProperty("VectorLength", VectorLength);
+        databaseJson.addProperty("UserRecords", serializedUserRecords);
+
+        // Write json object to file
+        String databaseString = databaseJson.toString();
+        try(FileOutputStream fos = new FileOutputStream(Filepath)) {
+            fos.write(databaseString.getBytes());
+        } catch (IOException e) {
+            Log.e(LogTag, "Cannot save database");
+            e.printStackTrace();
+            return;
+        }
+
+        Log.d(LogTag, "Database file successfully saved");
+    }
+
+    /**
+     * Temporary function to test database functionality.
+     *
+     */
+    public void generateDatabase() {
+        // Erase current database
+        usersRecords.clear();
+
+        // Generate sample vector
+        float[] sampleVector = new float[VectorLength];
+        Random r = new Random();
+        for(int i = 0; i < VectorLength; i++){
+            sampleVector[i] = r.nextFloat();
+        }
+
+        // Add sample data
+        usersRecords.put("Kuba", sampleVector);
+        usersRecords.put("Maciej", sampleVector);
+        usersRecords.put("Dominik", sampleVector);
+
+        // Save database in internal memory
+        saveDatabase();
+    }
+}

```

Comment:
IDK where to place it, but there is missing function to add another vector to same user and calculate average of them.

Reply:
I have thought about it, but it might be added later.

Reply:
+1